### PR TITLE
Group Contact create -> fix to use post hook for subscription records

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Update GroupContact method to create Subscription records in the post hook

Before
----------------------------------------
Done in the create method

After
----------------------------------------
Done in a post hook

Technical Details
----------------------------------------
Some complexities in getting the tests to pass mean this was left out when @colemanw did the others - should be passing now

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Contact/BAO/GroupContact.php [hunk 1]
```diff
@@ -9,66 +9,54 @@
...
-    return (!empty($params['id']) || (!empty($params['group_id']) && !empty($params['contact_id'])));
+  public static function self_hook_civicrm_post(PostEvent $event): void {
+    if (is_object($event->object) && in_array($event->action, ['create', 'edit'], TRUE)) {
+      // Lookup existing info for the sake of subscription history
+      if ($event->action === 'edit') {
+        $event->object->find(TRUE);
+      }
+      $params = $event->object->toArray();
+      try {
+        SubscriptionHistory::save(FALSE)->setRecords([$params])->execute();
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/GroupContact.php:52]: @colemanw I was worried about whether the subscription is created so I added test cover for it - in the process I added the SubscriptionHistory entity so as part of that I figured I might as well call that from the hook here - however, it writes and then gets rolled back - I'm a bit mystified as to why

**Comment by eileenmcnaughton** [comment]
Test fail is Stacktrace
api_v3_GroupContactTest::testDeleteWithPending with data set "APIv4" (4)
Failed asserting that actual size 0 matches expected size 1.

/home/jenkins/bknix-dfl/build/core-22419-55dv3/web/sites/all/modules/civicrm/tests/phpunit/api/v3/GroupContactTest.php:237
/home/jenkins/bknix-dfl/build/core-22419-55dv3/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:262
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671 

per comment above - I can fix by not calling the api & just calling writeRecords - but I'd like to understand why - since I've seen unexplained api rollbacks before & never gotten to the bottom

**Comment by colemanw** [comment]
@eileenmcnaughton we can start by adding the entity #22435

**Comment by eileenmcnaughton** [comment]
thanks @colemanw -regarding unexpected rollbacks (as replicated in the unit test here) - that seems to be an active discussion on chat too -

**Comment by colemanw** [comment]
Where on chat @eileenmcnaughton ?

**Comment by eileenmcnaughton** [comment]
@colemanw there is some concern about rollbacks in IATS - I haven't dug in at all though - but in this case - the record is being written to the table in the post hook but NOT being committed  -

**Comment by eileenmcnaughton** [comment]
@colemanw this is passing now ....

File: CRM/Contact/BAO/GroupContact.php [hunk 2]
```diff
@@ -9,59 +9,63 @@
...
+use Civi\Api4\SubscriptionHistory;
+use Civi\Core\Event\PostEvent;
+
 /**
  *
  * @package CRM
  * @copyright CiviCRM LLC https://civicrm.org/licensing
  */
-class CRM_Contact_BAO_GroupContact extends CRM_Contact_DAO_GroupContact {
+class CRM_Contact_BAO_GroupContact extends CRM_Contact_DAO_GroupContact implements \Civi\Test\HookInterface {
```
Comment (colemanw) [inline CRM/Contact/BAO/GroupContact.php:20]:
```suggestion
class CRM_Contact_BAO_GroupContact extends CRM_Contact_DAO_GroupContact implements \Civi\Core\HookInterface {
```

Comment (colemanw) [reply CRM/Contact/BAO/GroupContact.php:20]: @eileenmcnaughton this was renamed since Jan

Comment (eileenmcnaughton) [reply CRM/Contact/BAO/GroupContact.php]: @colemanw OK - fixed

**Pull request merged**